### ISSUE TITLE
Notable frame drops between version 28.98 and the current version.

### ISSUE THREAD

**Issue by AgedSage** [McKay42/McOsu#209]
I updated my game after not having done so for several months, after updating I noticed that the game did not feel nearly as smooth. I downpatched the game to 28.98(the version I used to use) and found that the framerate was perfect. I recorded the difference with these two videos. 
Both of these use the exact same configuration files.
Sorry I forgot to remove the handcam, I just used the setup I normally use if I want to stream.
Old version(28.98):
https://youtu.be/cdqeFGztXRw
Newest version:
https://youtu.be/FIdVHuF1F1A
You can see that the drops are very noticable in the second video, while the framerate never goes below 400 in the first video.

**Comment by McKay42** [comment]
Thanks for the videos, helps a lot compared to just text.
I've looked over the code changes since 28.98 (and all intermediary releases), and the only changes that could potentially affect frame pacing like this are these:
- Audio panning (depending on hitobject position, hitsounds are panned from left to right)
- WM_PAINT handling (I switched from BeginPaint()/EndPaint() to ValidateRect())

There are some other differences, but none which would cause random lag spikes which last a few frames. Most of the changes affect everything equally, but you have these random symptoms, which suggests some kind of outside influence (i.e. not in the raw game loop logic, but caused by thread context switches or library function calls like BASS audio).

Can you please try setting the affinity of McEngine.exe in Task Manager to only the last CPU core you have?
Like this: https://imgur.com/a/5dMyiKD
Check if it makes a difference.

**Comment by AgedSage** [comment]
Okay, I have tried that. 
I did not "feel" any more of the hitches, but I might be bit too tired to play well enough to notice at the moment.
When I looked at a video I recorded with autopilot the framerate seemed much more stable, though it was obviously lower.
Here's the video:
https://youtu.be/yylUxmDRRak
The largest drop was at about 1:20, but I'm not sure if that is unusual at all with only using one core.

**Comment by McKay42** [comment]
What's your processor (and all other specs)? The game is singlethreaded by design, so switching to one core shouldn't decrease your fps so much overall. Can you try setting the affinity to only the first core (maybe they are not equal)?

Also, these drops no longer look like spikes, it just drops for a few seconds consistently, therefore it can only be caused by the environment (everything else running on your computer). You can partially confirm this by pausing the game and checking the fps counter for some time (if paused, the game is doing absolutely nothing except rendering the static scene), if it still varies wildly then something else is causing that.

**Comment by AgedSage** [comment]
Sure, I'm using a 6600k(not overclocked), a 980ti, and 16GB of 2133MHz memory.
I'd be really surprised if it was the environment, those two videos I first posted were made right after each other, and I don't really have any taxing programs running on my computer.
Despite that I didn't really notice any more frame drops like that when I had the game use autopilot, but I also didn't see any frame drops when paused at all.
Here's a vid where I tried a few different settings, (with and without setting it to use one core, and with and without limiting the framerate, you can check the description to see which is which)
https://youtu.be/jmWfuPUB7N4
Despite not seeing any of the drops in this video, I could have sworn I felt a couple when I was playing without autopilot. So I'll try to record those too.
Also if it's worth mentioning, I am using NVENC encoding, so most of the processing involved in recording this should be on the graphics card, not the CPU.

**Comment by AgedSage** [comment]
It feels as smooth as ever now, although I'll have to put in a bit more time to make sure. I'm curious what you changed. Either way well done.
Also, is there a place where one can make feature requests? Do you just use the steam forums for that?

**Comment by McKay42** [comment]
Mostly Discord ([https://discord.gg/3UQnt7X](https://discord.gg/3UQnt7X)) and the Steam forums, yeah.

I'm not exactly sure how all of it relates to your specific symptoms, but I did fix an age-old brainfart in the delta time calculation of the main engine loop, which has reduced framerate jitter quite a bit. However, this jitter has nothing to do with your lag spikes, and is not something players usually notice since it averages out (not even myself, except after doing some frametime graph analysis now).

I also changed 2 (irrelevant?) BASS audio parameters for hitsound playback to match osu! (maximum number of simultaneous allowed samples, and sample override behavior), though again these should in theory not affect lag spikes whatsoever.

You'll have to use it a bit longer to make sure this actually made a difference.
